### ISSUE TITLE
MKDADensity does not work with FDRCorrector or FWECorrector(method='bonferroni')

### ISSUE THREAD

**Issue by jdkent** [neurostuff/NiMARE#348]
When trying to correct for multiple comparisons using FDR or FWE (using bonferroni, the montecarlo method works fine), the following error arises.

```
ValueError: <class 'nimare.correct.FDRCorrector'> requires "p" maps to be present in the MetaResult, but none were found.
```

psuedo-code
```
from nimare.correct import FDRCorrector
from nimare.meta import mkda

meta = mkda.MKDADensity()
corr = FDRCorrector()
res = meta.fit(dset)
cres = corr.transform(res)  # this is where the error occurs
```

**Comment by tsalo** [comment]
Unlike #347, I don't think this is an easy fix. The MKDA methods (minus MKDA Chi2) don't have any significance testing. The core algorithm just calculates the OF maps (essentially the MKDA equivalent to an ALE map). The significance testing happens with max-value Monte Carlo procedure in the `FWECorrector`, so there's no p-value map to correct for the other Correctors.

I will need to review both the original MKDA papers and possibly the ALE code for deriving its first null distribution to see if there's a native method that I missed or if we can adapt ALE's method to MKDA, respectively.

**Comment by tyarkoni** [comment]
@tsalo in that case, maybe we should separate permutation generation from monte carlo correction? In principle, I think one should be able to generate permutation-based p-value maps independently of what correction procedure one wants to use. I.e., I don't see a principled reason why one couldn't generate p-values for MKDA via permutation, and then apply bonferroni correction to that.

**Comment by tsalo** [comment]
This is kind of a naive question, but wouldn't the permutation-based p-value maps still be FWE-corrected? The procedure is basically the same as the ALE Monte Carlo FWE-correction procedure, after all.

**Comment by tyarkoni** [comment]
I'd have to look at the implementation, but the point is just that the generation of permutations is independent of the correction procedure. In the FWE procedure, you generate a bunch of permutations, take the max statistic in each permutation, and then select a threshold based on the Nth percentile of the distribution of max stats. This is the gold standard approach, and should probably remain the default/recommended correction procedure. But you don't *have* to do it this way; you could, e.g., just construct the null distribution of test statistics at each voxel in the brain, and then refer the observed value to that distribution to get a p-value at each voxel—at which point you can theoretically apply most of the standard MC correction procedures (e.g., pick an arbitrary threshold like p < .0001; use Bonferroni, etc.).

Since the computationally intensive part of the procedure is (I'm assuming) the actual permutation of foci/clusters, it would make sense to abstract that part out of the FWE correction procedure, and have the latter invoke the permutation generation implicitly before taking the max statistic. 

If you don't want to mess with the existing FWE correction implementation, that's totally cool, but we would still want to add standalone permutation generation routines, so that we can get a permutation-based p-value at each voxel. At that point, we can address the issue @jdkent flagged—i.e., we would be able to FDR or Bonferroni-correct the permutation-based p-values, just like any other p-values.

Basically what I'm saying is that, for the major estimators, there are in my view really three separate choices the user is making: (1) the estimator, (2) how the p-values are generated (analytically vs. permutation), and (3) how to correct those p-values. It may be that not all possible combinations are going to be valid, but I do think we should be supporting the ones in the title.

**Comment by tsalo** [comment]
In terms of actual usage, would it then look something like this?

```python
from nimare.correct import FWECorrector
from nimare.meta import mkda, ale

meta = mkda.MKDADensity(n_iters=10000, n_cores=2)
corr = FWECorrector(method="bonferroni")  # "montecarlo" no longer available
res = meta.fit(dset)
cres = corr.transform(res)

meta2 = ale.ALE()
corr2 = FWECorrector(method="montecarlo", n_iters=10000, n_cores=2)
res2 = meta2.fit(dset)
cres2 = corr2.transform(res2)  # same basic procedure as MKDADensity.fit()
```
